File: commands/stats/clanprofile.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
from datetime import datetime, timezone
from collections import Counter
import re
from database import db

class ClanProfile(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @app_commands.command(name='clanprofile', description='Detailed clan profile card with all statistics')
    async def clanprofile(self, interaction: discord.Interaction, clan: str):
        try:
            await interaction.response.defer()
            
            clan_upper = clan.upper().strip()
            query = {'clan_name': {'$regex': f'^{re.escape(clan_upper)}$', '$options': 'i'}}
            
            pipeline = [
                {'$match': query},
                {'$facet': {
                    'stats': [{'$group': {
                        '_id': None,
                        'total': {'$sum': 1},
                        'contests': {'$sum': {'$cond': ['$contest', 1, 0]}},
                        'points': {'$sum': {'$cond': ['$contest', {'$multiply': ['$player_count', 2]}, '$player_count']}},
                        'total_gain': {'$sum': '$gain'}
                    }}],
                    'maps': [{'$group': {'_id': '$map', 'count': {'$sum': 1}}}, {'$sort': {'count': -1}}, {'$limit': 5}],
                    'players': [{'$unwind': '$clan_winners'}, {'$group': {'_id': '$clan_winners', 'count': {'$sum': 1}}}, {'$sort': {'count': -1}}, {'$limit': 5}],
                    'recent': [{'$sort': {'timestamp': -1}}, {'$limit': 5}, {'$project': {'map': 1, 'time': 1}}]
                }}
            ]
            
            result = await db.db.clanwins.aggregate(pipeline).to_list(1)
            if not result or not result[0].get('stats'):
                await interaction.followup.send(embed=discord.Embed(
                    title=f"[{clan_upper}] Profile", description="No data found.", color=0xff0000
                ))
                return
            
            data = result[0]
            stats = data['stats'][0] if data['stats'] else {}
            total_wins = stats.get('total', 0)
            contest_wins = stats.get('contests', 0)
            total_points = stats.get('points', 0)
            total_gain = stats.get('total_gain', 0)
            avg_gain = total_gain / total_wins if total_wins > 0 else 0
            
            top_maps = [(m['_id'], m['count']) for m in data.get('maps', [])]
            top_map = top_maps[0] if top_maps else ('Unknown', 0)
            
            players_data = data.get('players', [])
            unique_players = len(players_data)
            top_player = (players_data[0]['_id'], players_data[0]['count']) if players_data else ('Unknown', 0)
            
            recent_wins = data.get('recent', [])
            last_win = recent_wins[0].get('time', 'Unknown')[:16] if recent_wins else 'Unknown'
            
            embed = discord.Embed(
                title=f"[{clan_upper}]",
                description="Complete Clan Profile",
                color=0x5865F2
            )
            
            overview = f"**Wins:** {total_wins}\n"
            overview += f"**Contest:** {contest_wins} ({contest_wins/total_wins*100:.1f}%)\n" if total_wins > 0 else "**Contest:** 0\n"
            overview += f"**Non-Contest:** {total_wins - contest_wins}\n"
            overview += f"**Points:** {total_points}\n"
            overview += f"**Gain:** {total_gain:.5f}"
            embed.add_field(name="Overview", value=overview, inline=False)
            
            performance = f"**Favorite Map:** {top_map[0]} ({top_map[1]})\n"
            performance += f"**Avg Gain/Win:** {avg_gain:.5f}\n"
            performance += f"**Contest Rate:** {contest_wins/total_wins*100:.1f}%" if total_wins > 0 else "**Contest Rate:** 0%"
            embed.add_field(name="Performance", value=performance, inline=True)
            
            team = f"**Unique Players:** {unique_players}\n"
            team += f"**MVP:** {top_player[0]} ({top_player[1]} wins)\n"
            team += f"**Last Win:** {last_win}"
            embed.add_field(name="Team", value=team, inline=True)
            
            maps_text = '\n'.join([f"{i+1}. {name} - {count}" for i, (name, count) in enumerate(top_maps)])
            embed.add_field(name="Top 5 Maps", value=maps_text, inline=True)
            
            pipeline_players = [
                {'$match': query},
                {'$unwind': '$clan_winners'},
                {'$group': {
                    '_id': '$clan_winners',
                    'count': {'$sum': 1},
                    'points': {'$sum': {'$cond': ['$contest', {'$multiply': ['$player_count', 2]}, '$player_count']}}
                }},
                {'$sort': {'count': -1}},
                {'$limit': 5}
            ]
            top_players_full = await db.db.clanwins.aggregate(pipeline_players).to_list(5)
            players_text = [f"{i+1}. {p['_id']} - {p['count']} ({p['points']}pts)" for i, p in enumerate(top_players_full)]
            embed.add_field(name="Top 5 Players", value='\n'.join(players_text), inline=True)
            
            recent_text = '\n'.join([f"{win.get('map', 'Unknown')}" for win in recent_wins])
            embed.add_field(name="Recent Wins", value=recent_text, inline=True)
            
            total_games = await db.db.clanwins.count_documents({})
            embed.set_footer(text=f"Data from 18 Nov 2025 | {total_games:,} games tracked")
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            print(f"Error in clanprofile: {e}")
            try:
                await interaction.followup.send(embed=discord.Embed(
                    title="Error", description="Failed to generate profile.", color=0xff0000
                ))
            except:
                pass

async def setup(bot):
    await bot.add_cog(ClanProfile(bot))

File: commands/stats/clanprofile.js
// https://github.com/viktorexe/terristats-discord-bot
// /clanprofile - a detailed clan profile card with overview, performance,
// team and recent-win sections.
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../../database');
const { exactMatch, gamesFooter } = require('../../utils');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('clanprofile')
    .setDescription('Detailed clan profile card with all statistics')
    .addStringOption((o) =>
      o.setName('clan').setDescription('Clan name').setRequired(true),
    ),

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const clan = interaction.options.getString('clan');
      const clanUpper = clan.toUpperCase().trim();
      const query = { clan_name: exactMatch(clanUpper) };

      const pipeline = [
        { $match: query },
        {
          $facet: {
            stats: [
              {
                $group: {
                  _id: null,
                  total: { $sum: 1 },
                  contests: { $sum: { $cond: ['$contest', 1, 0] } },
                  points: {
                    $sum: {
                      $cond: [
                        '$contest',
                        { $multiply: ['$player_count', 2] },
                        '$player_count',
                      ],
                    },
                  },
                  total_gain: { $sum: '$gain' },
                },
              },
            ],
            maps: [
              { $group: { _id: '$map', count: { $sum: 1 } } },
              { $sort: { count: -1 } },
              { $limit: 5 },
            ],
            players: [
              { $unwind: '$clan_winners' },
              { $group: { _id: '$clan_winners', count: { $sum: 1 } } },
              { $sort: { count: -1 } },
              { $limit: 5 },
            ],
            recent: [
              { $sort: { timestamp: -1 } },
              { $limit: 5 },
              { $project: { map: 1, time: 1 } },
            ],
          },
        },
      ];

      const result = await db.collection('clanwins').aggregate(pipeline).toArray();
      if (!result.length || !result[0].stats.length) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle(`[${clanUpper}] Profile`)
              .setDescription('No data found.')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const data = result[0];
      const stats = data.stats[0] || {};
      const totalWins = stats.total || 0;
      const contestWins = stats.contests || 0;
      const totalPoints = stats.points || 0;
      const totalGain = stats.total_gain || 0;
      const avgGain = totalWins > 0 ? totalGain / totalWins : 0;

      const topMaps = data.maps.map((m) => [m._id, m.count]);
      const topMap = topMaps.length ? topMaps[0] : ['Unknown', 0];

      const playersData = data.players || [];
      const uniquePlayers = playersData.length;
      const topPlayer = playersData.length
        ? [playersData[0]._id, playersData[0].count]
        : ['Unknown', 0];

      const recentWins = data.recent || [];
      const lastWin = recentWins.length
        ? (recentWins[0].time || 'Unknown').slice(0, 16)
        : 'Unknown';

      const embed = new EmbedBuilder()
        .setTitle(`[${clanUpper}]`)
        .setDescription('Complete Clan Profile')
        .setColor(0x5865f2);

      let overview = `**Wins:** ${totalWins}\n`;
      overview +=
        totalWins > 0
          ? `**Contest:** ${contestWins} (${((contestWins / totalWins) * 100).toFixed(1)}%)\n`
          : '**Contest:** 0\n';
      overview += `**Non-Contest:** ${totalWins - contestWins}\n`;
      overview += `**Points:** ${totalPoints}\n`;
      overview += `**Gain:** ${totalGain.toFixed(5)}`;
      embed.addFields({ name: 'Overview', value: overview, inline: false });

      let performance = `**Favorite Map:** ${topMap[0]} (${topMap[1]})\n`;
      performance += `**Avg Gain/Win:** ${avgGain.toFixed(5)}\n`;
      performance +=
        totalWins > 0
          ? `**Contest Rate:** ${((contestWins / totalWins) * 100).toFixed(1)}%`
          : '**Contest Rate:** 0%';
      embed.addFields({ name: 'Performance', value: performance, inline: true });

      let team = `**Unique Players:** ${uniquePlayers}\n`;
      team += `**MVP:** ${topPlayer[0]} (${topPlayer[1]} wins)\n`;
      team += `**Last Win:** ${lastWin}`;
      embed.addFields({ name: 'Team', value: team, inline: true });

      const mapsText = topMaps
        .map(([name, count], i) => `${i + 1}. ${name} - ${count}`)
        .join('\n');
      embed.addFields({ name: 'Top 5 Maps', value: mapsText, inline: true });

      const pipelinePlayers = [
        { $match: query },
        { $unwind: '$clan_winners' },
        {
          $group: {
            _id: '$clan_winners',
            count: { $sum: 1 },
            points: {
              $sum: {
                $cond: [
                  '$contest',
                  { $multiply: ['$player_count', 2] },
                  '$player_count',
                ],
              },
            },
          },
        },
        { $sort: { count: -1 } },
        { $limit: 5 },
      ];
      const topPlayersFull = await db
        .collection('clanwins')
        .aggregate(pipelinePlayers)
        .toArray();
      const playersText = topPlayersFull
        .map((p, i) => `${i + 1}. ${p._id} - ${p.count} (${p.points}pts)`)
        .join('\n');
      embed.addFields({ name: 'Top 5 Players', value: playersText, inline: true });

      const recentText = recentWins.map((w) => `${w.map || 'Unknown'}`).join('\n');
      embed.addFields({ name: 'Recent Wins', value: recentText, inline: true });

      const totalGames = await db.collection('clanwins').countDocuments({});
      embed.setFooter({ text: gamesFooter(totalGames) });
      await interaction.editReply({ embeds: [embed] });
    } catch (err) {
      console.log(`Error in clanprofile: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to generate profile.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
